File: Trie/208_implement_trie/208.cpp
#include <vector>
#include <string>

using namespace std;

struct TrieNode {
    vector<TrieNode*> nextNodes;
    bool end;
    TrieNode() : end(false), nextNodes(26, nullptr) {}
};

class Trie {

private:
    
    int index(char character) {
        return int(character) - 97;
    }
    
    void set_end(TrieNode* node) {
        node -> end = true;
    }
    
    void add(TrieNode* node, char character, TrieNode* newNode) {
        node -> nextNodes[index(character)] = newNode;
    }
    
    TrieNode* get(TrieNode* node, char character) {
        return node -> nextNodes[index(character)];
    }
    
    TrieNode* create() {
        TrieNode* node = new TrieNode();
        return node;
    }
    
    void insertRecur(TrieNode* node, int index, int &wordLen, string &word) {
        if (index == wordLen) {
            return set_end(node);
        }
        if (get(node, word[index]) == nullptr) {
            add(node, word[index], create());
        }
        return insertRecur(get(node, word[index]), index + 1, wordLen, word);
    }
    
    bool searchRecur(TrieNode* node, int index, int &wordLen, string &word) {
        if (index == wordLen) {
            return node -> end;
        }
        if (get(node, word[index]) == nullptr) {
            return false;
        }
        return searchRecur(get(node, word[index]), index + 1, wordLen, word);
    }
    
    bool startsWithRecur(TrieNode* node,int index,int &wordLen, string &word) {
        if (index == wordLen) {
            return true;
        }
        if (get(node, word[index]) == nullptr) {
            return false;
        }
        return startsWithRecur(get(node, word[index]), index + 1, wordLen, word);
    }
    
public:
    
    TrieNode* root;
    
    Trie() {
        root = new TrieNode();
    }
    
    void insert(string word) {
        int wordLen = word.size();
        return insertRecur(root, 0, wordLen, word);
    }
    
    bool search(string word) {
        int wordLen = word.size();
        return searchRecur(root, 0, wordLen, word);
    }
    
    bool startsWith(string prefix) {
        int wordLen = prefix.size();
        return startsWithRecur(root, 0, wordLen, prefix);
    }
};

File: Trie/208_implement_trie/208.py
# http://www.cse.yorku.ca/~oz/hash.html

class TrieNode:
    
    def __init__(self):
        # self.char = None
        self.next_nodes = [None] * 26
        self.end = False
        
    def index(self, char):
        return ord(char) - 97
    
    def set_end(self):
        self.end = True
    
    def add(self, char, node):
        self.next_nodes[self.index(char)] = node
        return
    
    def get(self, char):
        return self.next_nodes[self.index(char)]



class Trie:

    def __init__(self):
        self.root = TrieNode()

    def insert(self, word: str) -> None:
        def recur(node, idx):
            if idx == len(word):
                node.set_end()
                return 
            if node.get(word[idx]) == None:
                node.add(word[idx], TrieNode())
            return recur(node.get(word[idx]), idx + 1)
        idx = 0
        return recur(self.root, idx)

    def search(self, word: str) -> bool:
        def recur(node, idx):
            if idx == len(word):
                return node.end 
            if node.get(word[idx]) == None:
                return False
            return recur(node.get(word[idx]), idx + 1)
        idx = 0
        return recur(self.root, idx)
            

    def startsWith(self, prefix: str) -> bool:
        def recur(node, idx):
            if idx == len(prefix):
                return True
            if node.get(prefix[idx]) == None:
                return False
            return recur(node.get(prefix[idx]), idx + 1)
        return recur(self.root, 0)
            


# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)